### ISSUE TITLE
Feature request: Allowing playout cap without --noponder

### ISSUE THREAD

**Issue by roy7** [gcp/leela-zero#582]
I discussed this with GCP ages ago but it didn't make sense to sort out making this work for Leela since it's such a niche feature, that would only maybe benefit bot operators (or heck, just me :) ). 

Currently if you use --playouts, you also need to use --noponder. I'd like to request one of the volunteers out there working on LZ sort out a way to remove that requirement.

The idea is that after making a move, some playouts (or visits if a visit cap is added) will be pruned because of the paths explored but not taken. Thus whatever cap we just hit before moving, we're now under that cap and have some more work we could do while the opponent is thinking. Once we're back up to the cap, pondering stops and we just wait for the opponent's move.

Where this has come up before for me is running bots on OGS with various time settings and configurations. For ranked games for instance, with RoyalLeela, I run no playout cap and allow pondering. For unranked games and correspondence games I turn off pondering and have playout caps in place. The reason is because, unlike KGS where I believe bots only play 1 game at a time, OGS bots by default can play an unlimited number of games at once. Each game is fighting for the same GPU, so the more games playing at once, the weaker the bot will become.

If I want the bot to benefit from opponent time, I can't do a playout cap, and so the bot may think for long periods of time if the game setting is something like byoyomi 8h + 5m x 5. That's a long time for the GPU to be tied up on one game, while many other games might also be running. So a playout cap is needed... except with a playout cap and --noponder I can't start thinking until the opponent moves, which takes more time for the bot to make its move. The strength end result is the same (X playouts/visits are done before a move is made), but the user experience for the human opponent is better if the bot moves faster (sometimes all of the pondering work will be directly in the path the human went, so much of the playout/visit cap is still valid and new work done after opponent's move is very brief).

Another example of where bots might benefit from this is with correspondence games. Currently to prevent damage to RoyalLeela's strength in playing live ranked games, I have correspondence games not using the GPU at all. I also use a 100000 playout cap, since something like a Fischer 7 day + 1 day per move correspondence clock might let the bot think for 24 hours per move. With my Ryzen 1700 I don't mind if some CPU threads work on the game for a long time, but I don't want it to go for 20 hours. The advantage of pondering with a playout cap here is the same as above. We'll do as much work as we can during opponent's turn, then only do new work equal to the branches that get deleted when it is our turn. The same 100000 playouts are done, except the bot's resulting move back to the human might be far faster than otherwise, esp in forced move situations where nearly all of the search results during pondering will still be valid after opponent's move.

Anyway, this isn't overly important, doesn't directly impact the Leela Zero project, and wouldn't be needed for validation and selfplay matches. But it'd be a nice little improvement for at least a few of us, in case someone would care to tackle it eventually?

Months from now it's very possible Leela Zero ends up stronger than Leela itself, and Leela Zero becomes one of preeminent bots running on Go servers.

**Comment by Dorus** [comment]
Remember that most work during pondering will be lost. If the human player picks the preferred move, it might reuse 60-70% of the tree, but if a less favorable move is taken, that might be 1-10% only.

If you run multiple games simultaneous, it might be a good idea to lower the bot priority during pondering. That way it will only take up time pondering if no other games require attention.

Similar, you might want to run correspondence games with lower priority. 

That said, i'm not even sure if priority (on the cpu) works for the gpu. Or is gpy priority a thing too?

For pondering cap, it might be better to set the cap for pondering to the normal cap for each move. So if you have a pondering cap of 5000, every sub move will get 5000 playouts. You might want another upper cap for pondering too, else it end up with 5000 playouts on all 300 possible moves (wasting 99.7% of the calculations).

So much for my random ideas :-)

**Comment by Cabu** [comment]
I have created 2 pull requests with an implementation of this request:
 against main: https://github.com/gcp/leela-zero/pull/609
 against new: https://github.com/gcp/leela-zero/pull/610

**Comment by gcp** [comment]
I'm still at a loss at what the playout cap should do. I blocked this specifically because it cannot work in a sensible way.

A visit limit (#590) seems like it achieves what you want with the advantage that it actually works.

**Comment by roy7** [comment]
A playout or visit cap with pondering would allow faster response to a human move if things went as expected and prevent unlimited resource usage during pondering. Once visit counts are implemented I will move to those I think for bots.

**Comment by FFLaguna** [comment]
> A playout or visit cap with pondering would allow faster response to a human move if things went as expected and prevent unlimited resource usage during pondering. Once visit counts are implemented I will move to those I think for bots.

I've also wanted pondering + playout caps, for the exact same reason @roy7 mentions in the comment directly above mine.

**Comment by gcp** [comment]
Closing as --visits was just merged in the next branch.

**Comment by roy7** [comment]
Thanks! --visits may not truly be stronger, more tested is needed, but at least it's somewhat more intuitive perhaps.
